### PR TITLE
CFY-3035 - Added snapshot model.

### PR DESCRIPTION
**Pull request by xmarwas**

### CODE REVIEW

File: rest-service/manager_rest/resources.py [hunk 1]
```diff
@@ -728,24 +736,21 @@ def patch(self, execution_id):
...
 
 
-def get_snapshot_ctime(snapshot_path):
-    import time
-    return time.strftime('%d %b %Y %H:%M:%S',
-                         time.localtime(os.path.getctime(snapshot_path)))
+def _does_snapshot_exist(snapshot_id):
+    try:
+        get_blueprints_manager().get_snapshot(snapshot_id)
+    except:
```
Comment (simon-bar) [inline rest-service/manager_rest/resources.py]: Catch only manager_exceptions.NotFoundError here

File: rest-service/manager_rest/models.py [hunk 2]
```diff
@@ -40,6 +40,15 @@ def __init__(self, **kwargs):
         self.updated_at = kwargs['updated_at']
 
 
+class Snapshot(SerializableObject):
+
+    fields = {'id', 'created_at'}
+
+    def __init__(self, **kwargs):
+        self.id = kwargs['id']
```
Comment (simon-bar) [inline rest-service/manager_rest/models.py:54]: We also want a status here - like CREATING / UNDER_CONSTRUCTION, CREATED, FAILED, RESTORING (?)
We should update those fields appropriately in blueprint_manager's create_snapshot and restore_snapshot methods.

File: rest-service/manager_rest/resources.py [hunk 3]
```diff
@@ -782,11 +776,12 @@ class SnapshotsId(SecuredResource):
     @exceptions_handled
     @marshal_with(responses.Snapshot.resource_fields)
     def put(self, snapshot_id):
-        if _does_snapshot_exist(snapshot_id):
+        try:
+            snapshot = get_blueprints_manager().create_snapshot(snapshot_id)
+        except:
```
Comment (simon-bar) [inline rest-service/manager_rest/resources.py]: Same here, catch only the relevant exception

File: rest-service/manager_rest/resources.py [hunk 4]
```diff
@@ -782,11 +776,12 @@ class SnapshotsId(SecuredResource):
...
+            # we override original exeption
             raise RuntimeError("Snapshot with id '{0}' already exists."
```
Comment (simon-bar) [inline rest-service/manager_rest/resources.py]: Raise something from manager_exceptions, maybe BadParametersError, ConflictError or simply ManagerException?

File: rest-service/manager_rest/resources.py [hunk 5]
```diff
@@ -860,11 +843,16 @@ class SnapshotsIdUpload(SecuredResource):
     @exceptions_handled
     @marshal_with(responses.Snapshot.resource_fields)
     def put(self, snapshot_id):
+        if _does_snapshot_exist(snapshot_id):
+            raise RuntimeError("Snapshot with id '{0}' already exists."
```
Comment (simon-bar) [inline rest-service/manager_rest/resources.py]: Same here - "Raise something from manager_exceptions"

File: workflows/cloudify_system_workflows/snapshot.py [hunk 6]
```diff
@@ -44,17 +43,21 @@ def chunks(g):
...
         try:
             while s:
                 obj, idx = decoder.raw_decode(s)
+                n += 1
                 yield json.dumps(obj)
                 s = s[idx:]
         except:
             pass
 
+    assert not n or not s
```
Comment (simon-bar) [inline workflows/cloudify_system_workflows/snapshot.py:59]: Instead of assert we should probably raise some reasonable exception. Like NonRecoverableError for instance..

File: workflows/cloudify_system_workflows/snapshot.py [hunk 7]
```diff
@@ -44,17 +43,21 @@ def chunks(g):
             yield ch
 
     s = ''
+    n = 0
```
Comment (simon-bar) [inline workflows/cloudify_system_workflows/snapshot.py:46]: Maybe rename it to something like num_of_decoded_chunks.
This whole piece of code is likely to get replaced anyway soon..

File: rest-service/manager_rest/resources.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-#########
```
Comment (konradxyz) [inline rest-service/manager_rest/resources.py:1]: Kind of unnecessary change?

File: rest-service/manager_rest/resources.py (hunk 1 above)
Comment (xmarwas) [inline rest-service/manager_rest/resources.py]: This functions throws only RuntimeError
